server/core/src/main/java/dev/slimevr/tracking/trackers/hid/TrackersHID.kt

```diff
@@ -0,0 +1,249 @@
+	private val devicesByHID: MutableMap<HidDevice, MutableList<Int>> = HashMap()
+	private val hidServicesSpecification = HidServicesSpecification()
+	private val hidServices: HidServices
+
+	init {
+		hidServicesSpecification.isAutoStart = false
+		hidServices = HidManager.getHidServices(hidServicesSpecification)
+		hidServices.addHidServicesListener(this)
+		val dataReadThread = Thread(dataReadRunnable)
+		dataReadThread.isDaemon = true
+		dataReadThread.name = "hid4java data reader"
+		dataReadThread.start()
+	}
+
+	private fun checkConfigureDevice(hidDevice: HidDevice) {
+		if (hidDevice.vendorId == 0x2FE3 && hidDevice.productId == 0x5652) { // TODO: Use correct ids
+			if (!hidDevice.isOpen) {
+				check(hidDevice.open()) { "Unable to open device" }
+			}
+			// TODO: Configure the device here
+			val serial = hidDevice.serialNumber ?: "Unknown HID Device"
+			val product = hidDevice.product
+			val manufacturer = hidDevice.manufacturer
+			this.devicesBySerial[serial]?.let {
+				this.devicesByHID[hidDevice] = it
+				synchronized(this.devices) {
+					for (id in it) {
+						val device = this.devices[id]
+						for (value in device.trackers.values) {
+							if (value.status == TrackerStatus.DISCONNECTED) value.status = TrackerStatus.OK
+						}
+					}
+				}
+				LogManager.info("[TrackerServer] Linked HID device reattached: $serial")
+				return
+			}
+
+			// TODO: Need to check that the hidDevice is the same or different?
+			// TODO: Get firmware and manufacturer from the device?
+			val list: MutableList<Int> = mutableListOf()
+			this.devicesBySerial[serial] = list
+			this.devicesByHID[hidDevice] = list
+			LogManager.info("[TrackerServer] (Probably) Compatible HID device detected: $serial")
+		}
+	}
+
+	private fun removeDevice(hidDevice: HidDevice) {
+		this.devicesByHID[hidDevice]?.let {
+			synchronized(this.devices) {
+				for (id in it) {
+					val device = this.devices[id]
+					for (value in device.trackers.values) {
+						if (value.status == TrackerStatus.OK) {
+							value.status =
+								TrackerStatus.DISCONNECTED
+						}
+					}
+				}
+			}
+			this.devicesByHID.remove(hidDevice)
+			LogManager.info("[TrackerServer] Linked HID device removed: ${hidDevice.serialNumber}")
+		}
+	}
+
+	private fun setUpSensor(device: Device, trackerId: Int, sensorType: IMUType, sensorStatus: TrackerStatus) {
+		// LogManager.info("[TrackerServer] Sensor $trackerId for ${device.name}, status $sensorStatus")
+		var imuTracker = device.getTracker(trackerId)
+		if (imuTracker == null) {
+			imuTracker = Tracker(
+				device,
+				VRServer.getNextLocalTrackerId(),
+				device.name + "/" + trackerId,
+				"IMU Tracker #" + VRServer.currentLocalTrackerId,
+				null,
+				trackerNum = trackerId,
+				hasRotation = true,
+				hasAcceleration = true,
+				userEditable = true,
+				imuType = sensorType,
+				allowFiltering = true,
+				needsReset = true,
+				needsMounting = true,
+				usesTimeout = false
```

Comment:
using timeout is based on you executing the `heartbeat()` function of the tracker each time you receive a tracker packet.

The idea is that when it's taking too long it first gets a `TrackerStatus.TIMED_OUT` and after a few more it gets `TrackerStatus.DISCONNECTED`

The special thing about the `TIMED_OUT` status is to give less priority in tracker assignment and also to let the user know it's getting timed out.

It should be used in my opinion but the HID device should communicate some kind of heartbeat packet. (I know BLE is more like TCP, but maybe there is a way to have access to the heartbeat packets?)

Reply:
The timeout does get reset every time the tracker is updated, and it will work, but the tracker will not send updates if it doesn't need to. In this case, if timeout is used, whenever the tracker is updated the status also should be forced to `OK`.

The tracker also sleeps so quickly that heartbeat packets don't make sense. Plus, if the receiver is removed, all the linked trackers already get marked as disconnected, so I chose to not use timeout. Otherwise there might be no visible difference between tracker not sending data for any reason versus the receiver being disconnected.

Reply:
How do you know the device gets shutdown or lost connection then?

Reply:
In context of trackers themselves, there is no way, aside from using a timeout
Should timeout be used anyways, even if `DISCONNECTED` could have multiple causes?

Reply:
Ok so we go back again, how do you timeout a BLE tracker? What do you use as a heartbeat?